File: apps/fleet-manager-web/app/types/vehicle.ts
/**
 * 차량 관련 타입 정의
 */

// 차량 정보 타입
export interface Vehicle {
  id: string;
  vehicleId?: string;
  licensePlate?: string;
  plateNumber: string; // licensePlate의 별칭
  brand?: string;
  make: string; // brand의 별칭
  model: string;
  year: number;
  type?: 'sedan' | 'suv' | 'truck' | 'van';
  status:
    | 'active'
    | 'available'
    | 'in_use'
    | 'maintenance'
    | 'inactive'
    | 'reserved'
    | 'out_of_service'
    | 'idle';
  fuelType?: 'gasoline' | 'diesel' | 'electric' | 'hybrid';
  vinNumber?: string;
  vin?: string; // vinNumber의 별칭
  lastMaintenance?: string;
  nextMaintenance?: string;
  purchaseDate?: string;
  registrationDate?: string;
  insuranceExpiry?: string;
  imageUrl?: string;
  currentDriverId?: string;
  mileage?: number;
  fuelLevel?: number;
  lastInspection?: string;
  color?: string;
  location?: {
    latitude: number;
    longitude: number;
  };
  createdAt?: Date | string;
  updatedAt?: Date | string;
}

// 차량 위치 정보 타입
export interface VehicleLocation {
  id: string;
  vehicleId: string;
  latitude: number;
  longitude: number;
  timestamp: string;
}

// 차량 목록 응답 타입
export interface VehiclesResponse {
  vehicles?: Vehicle[]; // 기존 구조 유지
  items?: Vehicle[]; // 새로운 구조 지원
  total: number;
  page?: number;
  pageSize?: number;
  totalItems?: number;
  totalPages?: number;
  currentPage?: number;
  itemsPerPage?: number;
  pagination?: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

// 차량 필터 타입
export interface VehicleFilters {
  status?: string;
  type?: string;
  brand?: string;
  page?: number;
  pageSize?: number;
  search?: string;
}

// 차량 통계 타입
export interface VehicleStatistics {
  total: number;
  active: number;
  idle: number;
  maintenance: number;
  outOfService: number;
}

// 차량 폼 데이터 타입
export interface VehicleFormData {
  plateNumber: string;
  make: string;
  model: string;
  year: number;
  type?: 'sedan' | 'suv' | 'truck' | 'van';
  fuelType?: 'gasoline' | 'diesel' | 'electric' | 'hybrid';
  vinNumber?: string;
  color?: string;
  status?: Vehicle['status'];
  mileage?: number;
  fuelLevel?: number;
  notes?: string;
}

// 차량 정비 기록 타입
export interface VehicleMaintenance {
  id: string;
  vehicleId: string;
  type: string;
  date: string;
  notes: string;
  cost: number;
  serviceCenter: string;
  createdAt: string;
  updatedAt: string;
}

File: backend/shared/models/vehicle.py
"""
Vehicle 관련 모델 정의

차량 정보와 관련된 데이터 모델들을 정의합니다.
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field
from enum import Enum


class VehicleStatus(str, Enum):
    """차량 상태"""

    ACTIVE = "ACTIVE"
    INACTIVE = "INACTIVE"
    MAINTENANCE = "MAINTENANCE"
    RETIRED = "RETIRED"


class VehicleType(str, Enum):
    """차량 타입"""

    SEDAN = "SEDAN"
    SUV = "SUV"
    TRUCK = "TRUCK"
    VAN = "VAN"
    MOTORCYCLE = "MOTORCYCLE"
    OTHER = "OTHER"


class FuelType(str, Enum):
    """연료 타입"""

    GASOLINE = "GASOLINE"
    DIESEL = "DIESEL"
    ELECTRIC = "ELECTRIC"
    HYBRID = "HYBRID"
    LPG = "LPG"


class Vehicle(BaseModel):
    """차량 모델"""

    id: Optional[str] = None
    license_plate: str = Field(..., description="차량 번호판")
    user_id: Optional[str] = Field(None, description="소유자 ID")
    organization_id: Optional[str] = Field(None, description="조직 ID")

    # 차량 기본 정보
    make: Optional[str] = Field(None, description="제조사")
    model: Optional[str] = Field(None, description="모델명")
    year: Optional[int] = Field(None, description="연식")
    color: Optional[str] = Field(None, description="색상")
    vin: Optional[str] = Field(None, description="차대번호")

    # 차량 분류
    vehicle_type: Optional[VehicleType] = Field(None, description="차량 타입")
    fuel_type: Optional[FuelType] = Field(None, description="연료 타입")

    # 차량 상태
    status: VehicleStatus = Field(default=VehicleStatus.ACTIVE, description="차량 상태")
    mileage: Optional[int] = Field(None, description="주행거리 (km)")

    # 보험 및 등록 정보
    insurance_company: Optional[str] = Field(None, description="보험사")
    insurance_policy_number: Optional[str] = Field(None, description="보험증권번호")
    insurance_expiry_date: Optional[datetime] = Field(None, description="보험 만료일")
    registration_expiry_date: Optional[datetime] = Field(
        None, description="등록 만료일"
    )

    # 추가 정보
    notes: Optional[str] = Field(None, description="비고")
    tags: List[str] = Field(default_factory=list, description="태그")

    # 메타데이터
    created_at: Optional[datetime] = Field(None, description="생성일시")
    updated_at: Optional[datetime] = Field(None, description="수정일시")

    class Config:
        json_encoders = {datetime: lambda v: v.isoformat() if v else None}
        use_enum_values = True


class VehicleCreate(BaseModel):
    """차량 생성 요청 모델"""

    license_plate: str = Field(..., description="차량 번호판")
    user_id: Optional[str] = Field(None, description="소유자 ID")
    organization_id: Optional[str] = Field(None, description="조직 ID")
    make: Optional[str] = Field(None, description="제조사")
    model: Optional[str] = Field(None, description="모델명")
    year: Optional[int] = Field(None, description="연식")
    color: Optional[str] = Field(None, description="색상")
    vin: Optional[str] = Field(None, description="차대번호")
    vehicle_type: Optional[VehicleType] = Field(None, description="차량 타입")
    fuel_type: Optional[FuelType] = Field(None, description="연료 타입")
    mileage: Optional[int] = Field(None, description="주행거리 (km)")
    notes: Optional[str] = Field(None, description="비고")
    tags: List[str] = Field(default_factory=list, description="태그")


class VehicleUpdate(BaseModel):
    """차량 수정 요청 모델"""

    license_plate: Optional[str] = Field(None, description="차량 번호판")
    make: Optional[str] = Field(None, description="제조사")
    model: Optional[str] = Field(None, description="모델명")
    year: Optional[int] = Field(None, description="연식")
    color: Optional[str] = Field(None, description="색상")
    vin: Optional[str] = Field(None, description="차대번호")
    vehicle_type: Optional[VehicleType] = Field(None, description="차량 타입")
    fuel_type: Optional[FuelType] = Field(None, description="연료 타입")
    status: Optional[VehicleStatus] = Field(None, description="차량 상태")
    mileage: Optional[int] = Field(None, description="주행거리 (km)")
    insurance_company: Optional[str] = Field(None, description="보험사")
    insurance_policy_number: Optional[str] = Field(None, description="보험증권번호")
    insurance_expiry_date: Optional[datetime] = Field(None, description="보험 만료일")
    registration_expiry_date: Optional[datetime] = Field(
        None, description="등록 만료일"
    )
    notes: Optional[str] = Field(None, description="비고")
    tags: List[str] = Field(default_factory=list, description="태그")


class VehicleResponse(BaseModel):
    """차량 응답 모델"""

    id: str
    license_plate: str
    user_id: Optional[str] = None
    organization_id: Optional[str] = None
    make: Optional[str] = None
    model: Optional[str] = None
    year: Optional[int] = None
    color: Optional[str] = None
    vin: Optional[str] = None
    vehicle_type: Optional[VehicleType] = None
    fuel_type: Optional[FuelType] = None
    status: VehicleStatus
    mileage: Optional[int] = None
    insurance_company: Optional[str] = None
    insurance_policy_number: Optional[str] = None
    insurance_expiry_date: Optional[datetime] = None
    registration_expiry_date: Optional[datetime] = None
    notes: Optional[str] = None
    tags: List[str] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
